### PR TITLE
- rework on revive system

### LINKED ISSUE #73: Rework PPEffects in Medical System

### LINKED ISSUE #119: Bandage time for non medics seems weired
I found some really strange issues with the bandage speed when comparing medics with non medics.

### PR DESCRIPTION
**Pull request by netfusion**
This will restructure the revive module as well as the cfgRevive config

fixes #73
fixes #119

### CODE REVIEW

File: serverAddons/PRA3/addons/PRA3_Server/Revive/HitEffects/fn_blood.sqf [hunk 1]
```diff
@@ -0,0 +1,74 @@
...
+        PRA3_Player setDamage 1;
+    };
+}] call CFUNC(addPerFrameHandler);
+
+// Visual bleeding effect
+[{
+    private _bloodLoss = PRA3_Player getVariable [QGVAR(bloodLoss), 0];
+
+    if (_bloodLoss > 0) then {
+        [600 * _bloodLoss] call BIS_fnc_bloodEffect;
```
Comment (jokoho48) [inline serverAddons/PRA3/addons/PRA3_Server/Revive/HitEffects/fn_blood.sqf]: FUNC(bloodEffect) is our replacement for the BI effect

File: serverAddons/PRA3/addons/PRA3_Server/Revive/HitEffects/fn_clientInitHitEffectsUI.sqf [hunk 2]
```diff
@@ -0,0 +1,69 @@
+#include "macros.hpp"
```
Comment (jokoho48) [inline serverAddons/PRA3/addons/PRA3_Server/Revive/HitEffects/fn_clientInitHitEffectsUI.sqf]: wrong Filename? this render the Information if a Player is damaged and/or need revive or is bleeding

**Comment by jokoho48** [comment]
this PR may take serveal Weeks to test and to look whats changed

File: serverAddons/PRA3/addons/PRA3_Server/Revive/HitEffects/fn_unconscious.sqf [hunk 3]
```diff
@@ -0,0 +1,165 @@
...
+    PRA3_Player setVariable [QGVAR(isUnconscious), _state, true];
+    [_state] call CFUNC(disableUserInput);
+
+    if (_state) then {
+        if ((vehicle PRA3_Player) isKindOf "StaticWeapon") then {
+            moveOut PRA3_Player;
+        };
+
+        if ((animationState PRA3_Player) in ["ladderriflestatic", "laddercivilstatic"]) then {
+            PRA3_Player action ["ladderOff", nearestBuilding PRA3_Player];
```
Comment (jokoho48) [inline serverAddons/PRA3/addons/PRA3_Server/Revive/HitEffects/fn_unconscious.sqf]: i am not sure if a player can be on a vehicle or ladder after raggdoll need to check this

File: serverAddons/PRA3/addons/PRA3_Server/Revive/HitEffects/fn_unconscious.sqf [hunk 4]
```diff
@@ -0,0 +1,165 @@
...
+}] call CFUNC(addEventHandler);
+
+["UnconsciousnessChanged", {
+    (_this select 0) params ["_state"];
+
+    PRA3_Player setVariable [QGVAR(isUnconscious), _state, true];
+    [_state] call CFUNC(disableUserInput);
+
+    if (_state) then {
+        if ((vehicle PRA3_Player) isKindOf "StaticWeapon") then {
```
Comment (jokoho48) [inline serverAddons/PRA3/addons/PRA3_Server/Revive/HitEffects/fn_unconscious.sqf]: why only checking for staticWeapon?

Comment (jokoho48) [inline serverAddons/PRA3/addons/PRA3_Server/Revive/HitEffects/fn_unconscious.sqf]: why only checking StaticWeapon? we can not pull players out of Vehicles currently

File: serverAddons/PRA3/addons/PRA3_Server/Revive/HitEffects/fn_unconscious.sqf [hunk 5]
```diff
@@ -0,0 +1,165 @@
...
+        ([UIVAR(BleedOutProgress)] call bis_fnc_rscLayer) cutRsc [UIVAR(BleedOutProgress), "PLAIN", 0];
+        private _display = uiNamespace getVariable [UIVAR(BleedOutProgress), displayNull];
+
+        (_display displayCtrl 3003) ctrlSetStructuredText parseText "YOU ARE UNCONSCIOUS AND BLEEDING";
+        (_display displayCtrl 3004) ctrlSetStructuredText parseText "Wait for help or respawn ...";
+
+        [{
+            params ["_display", "_id"];
+
+            if (isNull _display) exitWith {
```
Comment (jokoho48) [inline serverAddons/PRA3/addons/PRA3_Server/Revive/HitEffects/fn_unconscious.sqf]: dont need to check every frame can be removed in L163

File: serverAddons/PRA3/addons/PRA3_Server/Revive/Treatments/fn_bandage.sqf [hunk 6]
```diff
@@ -0,0 +1,89 @@
...
+        if (_target getVariable [QGVAR(bloodLoss), 0] == 0 || _target getVariable [QGVAR(medicalActionRunning), ""] != "") exitWith {false};
+
+        [QGVAR(StartMedicalAction), ["BANDAGE", _target, "Bandaging %1 ..."]] call CFUNC(localEvent);
+
+        true
+    }];
+    (findDisplay 46) displayAddEventHandler ["MouseButtonUp", {
+        params ["_display", "_button"];
+
+        if (_button <= 1 && GVAR(medicalActionRunning) == "BANDAGE") exitWith {
```
Comment (jokoho48) [inline serverAddons/PRA3/addons/PRA3_Server/Revive/Treatments/fn_bandage.sqf]: why larget equalto? isnt it posible to have more than 1?

File: serverAddons/PRA3/addons/PRA3_Server/Revive/Treatments/fn_clientInitTreatments.sqf [hunk 7]
```diff
@@ -0,0 +1,148 @@
...
+    params ["_item"];
+
+    // Restore old action
+    [QGVAR(RestoreWeapon)] call CFUNC(localEvent);
+
+    // Move the weapon on back
+    PRA3_Player action ["SwitchWeapon", PRA3_Player, PRA3_Player, 99];
+
+    // Create a weapon holder and fill it with the dummy item
+    private _weaponHolder = createVehicle ["GroundWeaponHolder", [0, 0, 0], [], 0, "NONE"];
```
Comment (jokoho48) [inline serverAddons/PRA3/addons/PRA3_Server/Revive/Treatments/fn_clientInitTreatments.sqf]: can be replaced with createSimpleObject

File: serverAddons/PRA3/addons/PRA3_Server/Revive/Treatments/fn_clientInitTreatments.sqf [hunk 8]
```diff
@@ -0,0 +1,148 @@
...
+    // Remove the exit action if it exists.
+    private _restoreWeaponActionId = PRA3_Player getVariable [QGVAR(restoreWeaponAction), -1];
+    if (_restoreWeaponActionId > -1) then {
+        PRA3_Player removeAction _restoreWeaponActionId;
+    };
+}] call CFUNC(addEventHandler);
+
+// Reset values on death
+[QGVAR(Killed), {
+    (_this select 0) params ["_unit"];
```
Comment (jokoho48) [inline serverAddons/PRA3/addons/PRA3_Server/Revive/Treatments/fn_clientInitTreatments.sqf]: we may want to change the value on Uncon too?

File: serverAddons/PRA3/addons/PRA3_Server/Revive/Treatments/fn_clientInitTreatmentsUI.sqf [hunk 9]
```diff
@@ -0,0 +1,171 @@
...
+            case "FirstAidKit": {"bandage"};
+            case "Medikit": {"heal"};
+            default {"use %1 on"};
+        };
+
+        _text = "";
+        {
+            _x params ["_weaponName", "_condition"];
+
+            if (_fakeWeaponName == _weaponName) then {
```
Comment (jokoho48) [inline serverAddons/PRA3/addons/PRA3_Server/Revive/Treatments/fn_clientInitTreatmentsUI.sqf]: better use exitwith so the loop dont contiue

File: serverAddons/PRA3/addons/PRA3_Server/Revive/Treatments/fn_heal.sqf [hunk 10]
```diff
@@ -0,0 +1,110 @@
...
+    // Publish start position for progress
+    private _selectionDamage = PRA3_Player getVariable [QGVAR(selectionDamage), GVAR(selections) apply {0}];
+
+    private _selectionDamageSorted = +_selectionDamage;
+    _selectionDamageSorted sort false;
+    private _highestDamage = _selectionDamageSorted select 0;
+
+    PRA3_Player setVariable [QGVAR(treatmentProgress), 1 - (_highestDamage / _maxDamage), true];
+
+    // Publish time
```
Comment (jokoho48) [inline serverAddons/PRA3/addons/PRA3_Server/Revive/Treatments/fn_heal.sqf]: wrong line?

**Comment by jokoho48** [comment]
next time please smaller commits thanks...

**Comment by jokoho48** [comment]
one of my question is why you sometimes split the scripts very much and sometimes you collect stuff that have 0 connections?

**Comment by jokoho48** [comment]
this PR currently break the Revive system Completly
current bugs i found after 5 min testing
- [ ] unit dont bleed if they get Uncon(Bug) 
- [ ] the healing Player dont have a UI text if he try to revive (Bug)
- [ ] the Bleeding dont stop if you try Bandage? (Bug)
- [ ] Uncon PPEffect dont get rest on respawn (Bug)

File: serverAddons/PRA3/addons/PRA3_Server/Revive/Treatments/fn_bandage.sqf (hunk 6 above)
Comment (netfusion) [inline serverAddons/PRA3/addons/PRA3_Server/Revive/Treatments/fn_bandage.sqf]: Which line do you refer to?

File: serverAddons/PRA3/addons/PRA3_Server/Revive/Treatments/fn_clientInitTreatments.sqf (hunk 7 above)
Comment (netfusion) [inline serverAddons/PRA3/addons/PRA3_Server/Revive/Treatments/fn_clientInitTreatments.sqf]: What is createSimpleObject?

Comment (simon84) [inline serverAddons/PRA3/addons/PRA3_Server/Revive/Treatments/fn_clientInitTreatments.sqf]: It's a new bi command, to create only a 3D object: https://community.bistudio.com/wiki/createSimpleObject

**Comment by netfusion** [comment]
can't reproduce:
unit dont bleed if they get Uncon - this is working for me (unconscious timer gets decreased) (https://github.com/drakelinglabs/projectrealityarma3/blob/revive/serverAddons/PRA3/addons/PRA3_Server/Revive/HitEffects/fn_blood.sqf#L53)
the Bleeding dont stop if you try Bandage? - handled here (https://github.com/drakelinglabs/projectrealityarma3/blob/revive/serverAddons/PRA3/addons/PRA3_Server/Revive/HitEffects/fn_blood.sqf#L48)

File: serverAddons/PRA3/addons/PRA3_Server/Revive/Treatments/fn_clientInitTreatments.sqf (hunk 8 above)
Comment (netfusion) [inline serverAddons/PRA3/addons/PRA3_Server/Revive/Treatments/fn_clientInitTreatments.sqf]: currentWeaponChanged is triggered on uncon so there is no need to reset values

**Comment by jokoho48** [comment]
but this is only for local Player not for Other players
may connected to the other bug

**Comment by netfusion** [comment]
bleeding is always calculated locally
client know that they are bandaged/revived/healed because [this variable](https://github.com/drakelinglabs/projectrealityarma3/blob/revive/serverAddons/PRA3/addons/PRA3_Server/Revive/Treatments/fn_clientInitTreatments.sqf#L115) is set

**Comment by jokoho48** [comment]
currently the player dont have damage or bleeding after going uncon

**Comment by netfusion** [comment]
I keep my damage when going unconscious (http://puu.sh/oszgz/4c828bcc37.jpg)

**Comment by jokoho48** [comment]
ok i think i fixed the last bug and now i think we can merge this if @netfusion and @simon84 give there ok to this PR

**Comment by simon84** [comment]
Nice. I have no abilities to test it at the moment. Maybe I have time this evening.

**Pull request merged**